### Answer (a)

Define the weight vector ω as:

$$
\omega=\begin{pmatrix}
\omega_A \\
\omega_B \\
\omega_C \\
\omega_D
\end{pmatrix}
$$

Define the vector of returns µ as:
$$
\mu=\begin{pmatrix}
0.02 \\
0.07 \\
0.15 \\
0.20
\end{pmatrix}
$$

In [30]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from scipy.optimize import fsolve

In [31]:
mu = np.array([[0.02],
               [0.07],
               [0.15],
               [0.20]])

Define the standard deviation matrix S as:

$$
S=\begin{pmatrix}
0.05 & 0 & 0 & 0 \\
0 & 0.12 & 0 & 0 \\
0 & 0 & 0.17 & 0 \\
0 & 0 & 0 & 0.25 \\
\end{pmatrix}
$$

In [32]:
S = np.array([[0.05, 0, 0, 0],
              [0, 0.12, 0, 0],
              [0, 0, 0.17, 0],
              [0, 0, 0, 0.25]])

In [33]:
R = np.array([[1, 0.3, 0.3, 0.3],
              [0.3, 1, 0.6, 0.6],
              [0.3, 0.3, 1, 0.6],
              [0.3, 0.6, 0.6, 1]])

SRS = S @ R @ S
SRS

array([[0.0025 , 0.0018 , 0.00255, 0.00375],
       [0.0018 , 0.0144 , 0.01224, 0.018  ],
       [0.00255, 0.00612, 0.0289 , 0.0255 ],
       [0.00375, 0.018  , 0.0255 , 0.0625 ]])

In [40]:
SRS_inv = np.linalg.inv(SRS)
# SRS_inv = np.round(np.linalg.inv(SRS), 4)

SRS_inv

# tt = SRS @ SRS_inv
# tt

array([[458.37837838, -37.83783784, -15.26232114, -10.37837838],
       [-21.62162162, 106.60660661, -27.02702703, -18.37837838],
       [-26.709062  ,   9.14149444,  53.11886281, -22.7027027 ],
       [-10.37837838, -32.16216216, -12.97297297,  31.17837838]])

In [35]:
I = np.array([[1],
              [1],
              [1],
              [1]])
I_t = np.transpose(I)

A = I_t @ SRS_inv @ I
print(f'A: {A[0][0]:.4f}')

B = I_t @ SRS_inv @ mu
print(f'B: {B[0][0]:.4f}')

C = mu.T @ SRS_inv @ mu
print(f'C: {C[0][0]:.4f}')

D = A * C - B**2
print(f'D: {D[0][0]:.4f}')

A: 422.9929
B: 6.8180
C: 0.8902
D: 330.0684


### Answer (c)
$$
\sigma_{\pi}^2 = \omega'\sum\omega =
\begin{pmatrix}
0.7851 \\
0.0539 \\
0.1336 \\
0.0275 \\
\end{pmatrix}'
\begin{pmatrix}
0.0025 & 0.0018 & 0.00255 & 0.00375 \\
0.0018 & 0.0144 & 0.01224 & 0.018 \\
0.00255 & 0.01224 & 0.0289 & 0.0255 \\
0.00375 & 0.018 & 0.0255 & 0.0625 \\
\end{pmatrix}
\begin{pmatrix}
0.7851 \\
0.0539 \\
0.1336 \\
0.0275 \\
\end{pmatrix} 
= 0.003412
$$

In [42]:
w = np.array([[0.7851],
              [0.0539],
              [0.1336],
              [0.0275]])

print(w)
print(SRS)
variance = w.T @ SRS @ w
print(f'Variance: {variance}')
std_ev = np.sqrt(variance)
print(f'Standard Deviation: {std_ev}')

[[0.7851]
 [0.0539]
 [0.1336]
 [0.0275]]
[[0.0025  0.0018  0.00255 0.00375]
 [0.0018  0.0144  0.01224 0.018  ]
 [0.00255 0.00612 0.0289  0.0255 ]
 [0.00375 0.018   0.0255  0.0625 ]]
Variance: [[0.00336804]]
Standard Deviation: [[0.05803482]]
